File: 2020/julia/AOC/src/day14.jl
module Day14 
import AOC.Parsers

struct Mask
	orer::UInt64 
	ander::UInt64
	addresses::Vector{UInt64}
end

function Mask(str::AbstractString, part1::Bool)
	if part1
		orer = parse(UInt64, replace(str, 'X'=>'0'), base=2)
		ander = parse(UInt64, replace(str, 'X'=>'1'), base=2)
		Mask(orer, ander, [])
	else 
		orer = parse(UInt64, replace(str, 'X'=>'0'), base=2)
		float = parse(UInt64, replace(replace(str, '1'=>'0'), 'X'=>'1'), base=2)
		Mask(orer, float, expand_address_mask(float))
	end
end

Mask() = Mask(0, 0, [])

apply_mask(mask, value) = (value | mask.orer) & mask.ander

function expand_address_mask(float)
	addresses = UInt64[]
	count = 0
	notfloat = ~float
	baseitr = mod1(float, 2)
	while count <= float
		if (count & notfloat) == 0
			push!(addresses, count)
			count += baseitr
		else 
			count += count & notfloat
		end
	end
	addresses
end

decode_addresses(mask, value) = value | mask.ander ⊻ mask.ander .| mask.addresses .| mask.orer

struct Assignment
	target::UInt64
	value::UInt64
end

function Assignment(targetstr::AbstractString, valuestr::AbstractString)
	target = parse(UInt64, match(r"mem\[(\d+)\]", targetstr).captures[1])
	value = parse(UInt64, valuestr)
	Assignment(target, value)
end

mutable struct Program 
	mask::Mask 
	memory::Dict{UInt64, UInt64}
end

values(program::Program) = Base.values(program.memory)

Program() = Program(Mask(), Dict())

function execute_instruction!(program, inst; part1=true)
	splat = split(inst)
	if splat[1] == "mask"
		program.mask = Mask(splat[3], part1)
	else 
		assign = Assignment(splat[1], splat[3])
		if part1
			program.memory[assign.target] = apply_mask(program.mask, assign.value)
		else
			for address in decode_addresses(program.mask, assign.target)
				program.memory[address] = assign.value
			end
		end
	end
end

function part1(;data=Parsers.day14())
	program = Program()
	for instruction in data 
		execute_instruction!(program, instruction)
	end
	Int(sum(values(program)))
end

function part2(;data=Parsers.day14())
	program = Program()
	for instruction in data 
		execute_instruction!(program, instruction, part1=false)
	end
	Int(sum(values(program)))
end

end

File: 2020/fsharp/AOC/Day14.fs
module AOC.Day14 

open System
open System.Text.RegularExpressions

let parseTarget str = 
    let m = Regex.Match(str, "mem\[(\d+)\]")
    let cap = Array.tail [| for g in m.Groups -> g.Value |]
    int64 cap.[0]

let executeInstruction maskCreator addressDecoder valueAssigner (program, mask) (instruction: string)  =
    let splat = instruction.Split([|" = "|], StringSplitOptions.RemoveEmptyEntries)
    if splat.[0] = "mask" then 
        program, (maskCreator splat.[1])
    else
        let targets = parseTarget splat.[0] |> addressDecoder mask
        valueAssigner program targets mask (int64 splat.[1]), mask

let part1 data = 
    let assignValue memory addresses (ormask, andmask, _) value =
        let address = Seq.exactlyOne addresses 
        Map.add address ((value ||| ormask) &&& andmask) memory

    let decodeAddress _ target = [target]

    let createMask str = 
        let orer = str |> Seq.map (fun c -> match c with | 'X' -> '0' | _ -> c) |> String.Concat
        let ander = str |> Seq.map (fun c -> match c with | 'X' -> '1' | _ -> c) |> String.Concat
        (Convert.ToInt64(orer, 2)), (Convert.ToInt64(ander, 2)), Seq.empty
    
    let executor = executeInstruction createMask decodeAddress assignValue
    Seq.fold executor (Map.empty, (0L, 0L, Seq.empty)) data 
    |> fst
    |> Map.fold (fun state _ value -> state + value) 0L 

let countWithMask floater =
    let notfloat = ~~~floater
    let rem = floater % 2L
    let baseitr = if rem = 0L then floater else rem
    let rec counter count = 
        if count > floater then 
            None 
        elif (count &&& notfloat) = 0L then
            Some (count, (count + baseitr))
        else 
            counter (count + (count &&& notfloat))
    Seq.unfold counter 0L

let part2 data = 
    let assignValue memory addresses _ value = 
        Seq.fold (fun map address -> Map.add address value map) memory addresses

    let decodeAddress (ormask, xormask, offsets) target = 
        offsets |> Seq.map (fun x -> (((target ||| xormask) ^^^ xormask) ||| x) ||| ormask)

    let createMask str = 
        let orer = str |> Seq.map (fun c -> match c with | 'X' -> '0' | _ -> c) |> String.Concat
        let floater = str |> Seq.map (fun c -> match c with | '1' -> '0' | 'X' -> '1' | _ -> c) |> String.Concat
        let floatint = Convert.ToInt64(floater, 2)
        Convert.ToInt64(orer, 2), floatint, countWithMask floatint

    let executor = executeInstruction createMask decodeAddress assignValue
    Seq.fold executor (Map.empty, (0L, 0L, Seq.empty)) data 
    |> fst
    |> Map.fold (fun state _ value -> state + value) 0L 